clients/src/main/java/org/apache/kafka/common/security/kerberos/Login.java

```diff
@@ -0,0 +1,410 @@
+                            try {
+                                reLogin();
+                                break;
+                            } catch (LoginException le) {
+                                if (retry > 0) {
+                                    --retry;
+                                    // sleep for 10 seconds.
+                                    try {
+                                        Thread.sleep(10 * 1000);
+                                    } catch (InterruptedException e) {
+                                        log.error("Interrupted during login retry after LoginException:", le);
+                                        throw le;
+                                    }
+                                } else {
+                                    log.error("Could not refresh TGT for principal: " + principal + ".", le);
+                                }
+                            }
+                        }
+                    } catch (LoginException le) {
+                        log.error("Failed to refresh TGT: refresh thread exiting now.", le);
+                        break;
+                    }
+                }
+            }
+        });
+        t.setDaemon(true);
+    }
+
+    public void startThreadIfNeeded() {
+        // thread object 't' will be null if a refresh thread is not needed.
+        if (t != null) {
+            t.start();
+        }
+    }
+
+    public void shutdown() {
+        if ((t != null) && (t.isAlive())) {
+            t.interrupt();
+            try {
+                t.join();
+            } catch (InterruptedException e) {
+                log.warn("error while waiting for Login thread to shutdown: " + e);
+            }
+        }
+    }
+
+    public Subject subject() {
+        return subject;
+    }
+
+    public String loginContextName() {
+        return loginContextName;
+    }
+
+    private synchronized LoginContext login(final String loginContextName) throws LoginException {
+        if (loginContextName == null) {
+            throw new LoginException("loginContext name (JAAS file section header) was null. " +
+                    "Please check your java.security.login.auth.config (=" +
+                    System.getProperty("java.security.login.auth.config") +
+                    ") and your " + JaasUtils.LOGIN_CONTEXT_SERVER + "(=" +
+                    System.getProperty(JaasUtils.LOGIN_CONTEXT_CLIENT, "Client") + ")");
+        }
+
+        if (System.getProperty("java.security.auth.login.config") == null) {
+            throw new IllegalArgumentException("You must pass java.security.auth.login.config in secure mode.");
+        }
+
+        File configFile = new File(System.getProperty("java.security.auth.login.config"));
+        Configuration loginConf = null;
+        try {
+            loginConf = Configuration.getInstance("JavaLoginConfig", new URIParameter(configFile.toURI()));
+            Configuration.setConfiguration(loginConf);
```

Comment:
I am wondering how well this works when the broker is enabled to also authenticate to ZK through sasl. Will the global Configuration be set twice (once here and another time potentially in zookeeper client)? Will that affect the login logic? @fpj , do you know?